Fix remainder type promotion

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48668 Fix remainder type promotion**
* #48278 Fix `fmod` type promotion

Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

## BC-breaking tests:
[Nightly](https://www.internalfb.com/intern/testinfra/testconsole/testrun/6192449517872347/)
[This PR](https://www.internalfb.com/intern/testinfra/testconsole/testrun/3659174740128654/)
This PR doesn't introduce any new failure or timeout model.

Differential Revision: [D25869136](https://our.internmc.facebook.com/intern/diff/D25869136)

ejguan committed:
Fix remainder type promotion

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "[WIP] Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

houseroad commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -3803,13 +3803,19 @@ def forward(self, input, other):
         y = torch.randn(1, 2, 1)
         self.run_test(RemainderModel(), (x, y))
 
+    # Note: Integral Tensor remainder to floating-point scalar (not Tensor)
```
this is bc-breaking change. Can we still make this case working for the original test? 

Otherwise, existing pipeline might just break. cc: @spandantiwari, @BowenBao

BowenBao replied:
@ejguan could you try removing the casting at here

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/onnx/symbolic_opset9.py#L2598

and see if it works for the original test case? The was workaround of onnx export for remainder since it was not adhering to type promotion before.

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan commented on test/onnx/expect/TestOperators.test_remainder.expect:
```diff
@@ -3,41 +3,30 @@ producer_name: "pytorch"
-    op_type: "Cast"
-    attribute {
-      name: "to"
-      i: 1
-      type: INT
-    }
+    name: "Div_0"
+    op_type: "Div"
```
Hi @BowenBao 
I updated the expected IR of remainder, since the operator of `remainder` is changed. And the `test/onnx/test_operator.py  TestOperators.test_remainder` is passed now.

ejguan commented:
@mruberry 
Could you review this PR and #48278 , since the ONNX errors on both PRs have been mitigated.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
-        x.fmod_(zero)
-        self.assertTrue(torch.all(x.isnan()))
+    def test_fmod_remainder_by_zero_float(self, device, dtype):
+        fn_list = [(torch.fmod, lambda x, mod: x.fmod_(mod)),
```
This enumeration of functions is cool but I think it can be simplified to:

`(torch.Tensor.fmod_, torch.Tensor.remainder_)`

mruberry replied:
Also, in general we should tuples for iterables. Lists are OK when the iterable has to be mutable.

mruberry replied:
Note that the behavior of returning NaN when the denominator of fmod or remainder is zero is consistent with the C++ standard but inconsistent with NumPy. That's OK. We should prefer the C++ or Python standards where applicable.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
+            self.assertTrue(torch.all(fns[0](x, 0.0).isnan()))
+            self.assertTrue(torch.all(fns[0](x, zero).isnan()))
+            # out
+            out = torch.empty(0, device=device, dtype=dtype)
```
I don't think you need to test the out or inplace variants for this behavior.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
-                self.assertEqual(x_ == 4294967295, x >= 0)
-                self.assertEqual(x_ == -1, x < 0)
+    def test_fmod_remainder_by_zero_integral(self, device, dtype):
+        fn_list = [(torch.fmod, lambda x, mod: x.fmod_(mod)),
```
See comments above.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
+            x = make_tensor((10, 10), device=device, dtype=dtype, low=-9, high=9)
+            zero = torch.zeros_like(x)
+            # out
+            out = torch.empty(0, device=device, dtype=dtype)
```
Once again for special cases it shouldn't be necessary to test variant behavior like out=.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
+            # In-place
+            x_ = x.clone()
+            # RuntimeError on CPU
+            if device == 'cpu':
```
`if self.device_type == 'cpu'` or `if torch.device(device).type == 'cpu'`

mruberry replied:
Since so much of PyTorch's code is copy-pasted it's important to use the blessed patterns, even when another, possibly simpler pattern, will work. For example, if a test tried to do `if device == 'cuda':` that would fail, because device could be `cuda:0` or `cuda:1` when the test is run on a CUDA device. That's why it's important to check the device's type using `self.device_type` or constructing a device with `torch.device(device)` and then acquiring its type.

ejguan replied:
It makes a lot of sense!

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
-                x_.fmod_(zero)
-                self.assertTrue(torch.all(x_ == value))
+                if dtype == torch.int64:
+                    self.assertEqual(fns[0](x, zero) == 4294967295, x >= 0)
```
Would you elaborate on our integer CUDA behavior? That is, where are these numbers coming from?

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
-    def test_fmod(self, device, dtype):
+    def test_fmod_remainder(self, device, dtype):
         # Use numpy as reference
         def _reference_implementation(x, mod):
```
Once again, "reference_implementation" seems like a misnomer because this function actually performs the test.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,100 +1359,109 @@ def test_rdiv(self, device, dtype):
-            res = torch.fmod(x, mod)
-            common_dtype = torch.promote_types(res.dtype, exp.dtype)
-            exp = exp.to(common_dtype)
+                res = fns[0](x, mod)
```
Analogous comments about type handling here as in the fmod PR

mruberry replied:
It would be more readable if the tuple was unpacked as `fn, inplace, ref` immediately instead of referencing its parts with indices.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1471,44 +1480,6 @@ def _reference_implementation(x, mod):
 
-    @onlyCPU
-    @dtypes(torch.float, torch.long)
-    def test_remainder(self, device, dtype):
```
Is it really OK to just remove this test and the following one? The new tests do not seem to cover everything that it does.

ejguan replied:
I think the following tests cover:
1. dtype for float and long
2. Tensor remainder to scalar or scalar tensor (positive and negative)
3. Tensor remainder to another tensor (broadcasting)
The only thing left, which I will put into the new test cases, is negative divisor.

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,72 +1359,79 @@ def test_rdiv(self, device, dtype):
+            if device == 'cpu':
+                with self.assertRaisesRegex(RuntimeError, "ZeroDivisionError"):
+                    fn(x, zero)
+            # Different value for different dtype on CUDA:
```
This note is very helpful. In a future PR we should probably update the documentation for fmod and remainder to include this, too.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1359,72 +1359,79 @@ def test_rdiv(self, device, dtype):
+                exp = torch.from_numpy(exp)
+
+                res = fn(x, mod)
+                # Different Type promotion between numpy and torch for
```
To enforce consistent integer->float type promotion you can wrap the NumPy references with

https://github.com/pytorch/pytorch/blob/573f4aa3529371e9d30a0884d56eac0e6d3ddd71/torch/testing/_internal/common_methods_invocations.py#L284

Then you wouldn't have to deal with this

mruberry commented:
Hey @ejguan, this PR looks pretty good. I made a few inline comments which should be easy to deal with.

One last thing to do on this PR is to write a "BC-breaking Note:" above the PR summary explaining what the previous behavior was vs. what the new behavior is, and to label this PR as "BC-breaking." These notes help the release notes writers for each version of PyTorch. This BC-break is very minor, but we might as well do it for completeness. 

I reviewed our operator database for uses of fmod and remainder that might actually change the behavior of existing models and didn't find any, so we're probably OK. If you're curious about a more thorough BC-breaking change test, however, let me know and I can walk through the process.

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -116,8 +116,8 @@ void div_kernel(TensorIterator& iter) {
 void remainder_kernel(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cpu", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ false)) {
```
What happens if remainder is called with a bool argument?

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Will fix part of #47779
Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### Previous:
- Type is determined by the first input (Tensor). Seconds input is casted to the `dtype` of the first input to get the output.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.int32
```
- Bool Tensor is not supported
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.randint(10, [10, 10], dtype=torch.int32)
>>> y = torch.randn(10, 10, dtype=torch.float32)
>>> out = torch.remainder(x, y)
>>> out.dtype
torch.float32
```
- Bool Tensor is supported as inputs

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
- For two tensor arguments,
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float32
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
- Bool Tensor is **not supported**
```python
>>> a = b = torch.tensor([True])
>>> torch.remainder(a, b)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
>>> torch.remainder(a, 1)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
```
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float64
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the dtype of result is promoted.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```
- Bool Tensor is **supported** as inputs
```python
>>> a = b = torch.tensor(True)
>>> torch.remainder(a, b)
tensor(False)
>>> torch.remainder(a, 0.7)
tensor(0.3000)
```
Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -116,8 +116,8 @@ void div_kernel(TensorIterator& iter) {
 void remainder_kernel(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cpu", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ true)) {
```
Like with fmod, let's exclude bool for remainder, too. NumPy also promotes bool inputs to remainder.

mruberry commented on aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:
```diff
@@ -12,8 +12,8 @@
 void remainder_kernel_cuda(TensorIterator& iter) {
-  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
-    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+  if (isIntegralType(iter.common_dtype(), /*includeBool*/ true)) {
```
Remove bool here, too.

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -1354,63 +1354,70 @@ def test_rdiv(self, device, dtype):
+            for fn, inplace_fn, ref_fn in fns_list:
+                np_x = x.cpu().numpy()
+                np_mod = mod.cpu().numpy() if torch.is_tensor(mod) else mod
+                ref_fn = np_binary_ufunc_type_promotion_wrapper(ref_fn)
```
See fmod PR for a proposal simplifying this.

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
- For two tensor arguments,
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float32
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
- Bool Tensor is **not supported**
```python
>>> a = b = torch.tensor([True])
>>> torch.remainder(a, b)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
>>> torch.remainder(a, 1)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
```
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float64
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the dtype of result is promoted.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```
- Bool Tensor is **supported** as inputs
```python
>>> a = b = torch.tensor(True)
>>> torch.remainder(a, b)
tensor(False)
>>> torch.remainder(a, 0.7)
tensor(0.3000)
```
Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
- For two tensor arguments,
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float32
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
- Bool Tensor is **not supported**
```python
>>> a = b = torch.tensor([True])
>>> torch.remainder(a, b)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
>>> torch.remainder(a, 1)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
```
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float64
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the dtype of result is promoted.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```
- Bool Tensor is **supported** as inputs
```python
>>> a = b = torch.tensor(True)
>>> torch.remainder(a, b)
tensor(False)
>>> torch.remainder(a, 0.7)
tensor(0.3000)
```
Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
- For two tensor arguments,
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float32
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
- Bool Tensor is **not supported**
```python
>>> a = b = torch.tensor([True])
>>> torch.remainder(a, b)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
>>> torch.remainder(a, 1)
RuntimeError: "remainder_cpu" not implemented for 'Bool'
```
### Current:
- Type is determined by both inputs.
```python
>>> x = torch.arange(start=1, end=6, dtype=torch.int32)  # tensor([1, 2, 3, 4])
>>> y = torch.arange(start=1.1, end=2.1, step=0.2, dtype=torch.float32)  # tensor([1.1, 1.3, 1.5, 1.7, 1.9])
>>> torch.remainder(x, y)
tensor([1.0000, 0.7000, 0.0000, 0.6000, 1.2000])
>>> z = torch.arange(start=0.2, end=1.1, step=0.2, dtype=torch.float64)  # tensor([0.2, 0.4, 0.6, 0.8, 1.], dtype=torch.float64)
>>> torch.remainder(y, z).dtype
torch.float64
>>> torch.remainder(z, y).dtype
torch.float64
```
- For tensor and scalar arguments, the dtype of result is promoted.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```
- Bool Tensor is **supported** as inputs
```python
>>> a = b = torch.tensor(True)
>>> torch.remainder(a, b)
tensor(False)
>>> torch.remainder(a, 0.7)
tensor(0.3000)
```
Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

Differential Revision: [D25254460](https://our.internmc.facebook.com/intern/diff/D25254460)

[ghstack-poisoned]

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```


[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -128,7 +128,7 @@ void remainder_kernel(TensorIterator& iter) {
     });
   } else {
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "remainder_cpu", [&]() {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.common_dtype(), "remainder_cpu", [&]() {
```
Move this dispatch out of the else branch.

ejguan replied:
I think we still need to keep it within the else block. Otherwise, we will dispatch two times for this operation.

mruberry commented on aten/src/ATen/native/cuda/BinaryRemainderKernel.cu:
```diff
@@ -23,7 +23,7 @@ void remainder_kernel_cuda(TensorIterator& iter) {
       });
     });
   } else {
```
Move this dispatch out of the else branch.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -7306,17 +7306,17 @@ def merge_dicts(*dicts):
+.. note::
     Complex inputs are not supported. In some cases, it is not mathematically
     possible to satisfy the definition of a modulo operation with complex numbers.
+    Check the note from :func:`torch.fmod` for zero divisor.
```
"See :func:\`torch.fmod\` for how division by zero is handled."

mruberry reviewed (approved):
Nice work, @ejguan. Like with fmod, I've inlined a few simple changes to the kernels and a small docs update. Also like with fmod we should run an additional BC-breaking test on this diff before landing it.

ejguan committed:
Update on "Fix remainder type promotion"


Combine tests for `fmod` and `remainder`.

## BC-breaking Note:
In order to make `remainder` operator have type promotion, we have to introduce BC breaking.
### 1.7.1:
In the case where the second argument is a python number, the result is casted to the dtype of the first argument.
```python
>>> torch.remainder(x, 1.2)
tensor([0, 0, 0, 0, 0], dtype=torch.int32)
```
### This PR:
In the case where the second argument is a python number, the dtype of result is determined by type promotion of both inputs.
```python
>>> torch.remainder(x, 1.2)
tensor([1.0000, 0.8000, 0.6000, 0.4000, 0.2000])
```

Differential Revision: [D25869136](https://our.internmc.facebook.com/intern/diff/D25869136)

[ghstack-poisoned]
